src/mlpack/methods/reinforcement_learning/replay/prioritized_replay.hpp

```diff
@@ -0,0 +1,269 @@
+             bool isEnd)
+  {
+    states.col(position) = state.Encode();
+    actions(position) = action;
+    rewards(position) = reward;
+    nextStates.col(position) = nextState.Encode();
+    isTerminal(position) = isEnd;
+
+    idxSum.set(position, max_priority * alpha);
+
+    position++;
+    if (position == capacity)
+    {
+      full = true;
+      position = 0;
+    }
+  }
+
+  /**
+   * Sample some experience according to their priorities.
+   *
+   * @return The indices to be chosen.
+   */
+  arma::ucolvec sampleProportional()
+  {
+    arma::ucolvec idxes(batchSize);
+    double totalSum = idxSum.sum(0, (full ? capacity : position));
+    double sumPerRange = totalSum / batchSize;
+    for (size_t bt = 0; bt < batchSize; bt ++) {
+      double mass = arma::randu() * sumPerRange + bt * sumPerRange;
+      size_t idx = idxSum.findPrefixSum(mass);
+      idxes[bt] = idx;
+    }
+    return idxes;
+  }
+
+  /**
+   * Sample some experience according to their priorities.
+   *
+   * @param sampledStates Sampled encoded states.
+   * @param sampledActions Sampled actions.
+   * @param sampledRewards Sampled rewards.
+   * @param sampledNextStates Sampled encoded next states.
+   * @param isTerminal Indicate whether corresponding next state is terminal
+   *        state.
+   * @param sampledIndices Sampled indices
+   * @param weights Corresponding weight for updating the loss
+   */
+  void Sample(arma::mat& sampledStates,
+              arma::icolvec& sampledActions,
+              arma::colvec& sampledRewards,
+              arma::mat& sampledNextStates,
+              arma::icolvec& isTerminal,
+              arma::ucolvec& sampledIndices,
+              arma::rowvec& weights)
+  {
+    size_t upperBound = full ? capacity : position;
+
+    sampledIndices = sampleProportional();
+    betaAnneal();
+
+    sampledStates = states.cols(sampledIndices);
+    sampledActions = actions.elem(sampledIndices);
+    sampledRewards = rewards.elem(sampledIndices);
+    sampledNextStates = nextStates.cols(sampledIndices);
+    isTerminal = this->isTerminal.elem(sampledIndices);
+
+//  calculate the weights of sampled transitions
+
+    size_t num_sample = full ? capacity : position;
+    weights = arma::rowvec(sampledIndices.n_rows);
+
+    for (size_t i = 0; i < sampledIndices.n_rows; ++ i)
+    {
+      double p_sample = idxSum.get(sampledIndices[i]) / idxSum.sum();
+      weights[i] = pow(num_sample * p_sample, -beta);
+    }
+    weights /= weights.max();
+  }
+
+  /**
+   * Update priorities of sampled transitions.
+   *
+   * @param indices The indices of sample to be updated.
+   * @param priorities Their corresponding priorities.
+   */
+  void update_priorities(arma::ucolvec& indices, arma::colvec& priorities)
+  {
+      arma::colvec alphaPri(indices.n_rows);
+      for (size_t i = 0; i < indices.n_rows; ++i)
+      {
+        alphaPri = alpha * priorities[i];
+        max_priority = std::max(max_priority, priorities[i]);
```

Comment:
This is rather picky comment. But I guess we could replace this block with : 
```
alphaPri = alpha * priorities;
max_priority = std::max(max_priority, arma::max(alphaPri));
```
Can't say surely, whether it will make any difference in terms of speed. For that need to verify with armadillo `max` implementation.

Reply:
You are right, the vectorized code is much better. So I prefer using:
```
max_priority = std::max(max_priority, arma::max(alphaPri));
```
I think we can add a `TODO` tag for this block.